Source/WebCore/rendering/RenderBlockFlow.cpp

```diff
@@ -1367,6 +1394,55 @@ bool RenderBlockFlow::shouldTrimChildMargin(MarginTrimType marginTrimType, const
+        // since this is the float's positin within the BFC. If not, add the containing block's location
+        // to the float's location. This new value should be the location of the float relative to this
+        // new containing block. Keep doing this until we find the containing block that establishes
+        // a BFC
+        auto bfcRoot = floatBox.blockFormattingContextRoot();
+        ASSERT(bfcRoot);
+
+        auto startLocation = floatBox.localToAbsolute();
+        auto containingBlock = floatBox.containingBlock();
+        while (!containingBlock->establishesBlockFormattingContext()) {
+            startLocation += containingBlock->localToAbsolute();
+            containingBlock = containingBlock->containingBlock();
+        }
+        if (bfcRoot->isHorizontalWritingMode())
+            return LayoutUnit { startLocation.y() + bfcRoot->logicalHeightForChild(floatBox) };
+        return LayoutUnit { startLocation.x() + bfcRoot->logicalWidthForChild(floatBox) };
+    };
+    for (auto& floatingObject : m_floatingObjects->set()) {
+        auto& floatBox = floatingObject->renderer();
+        if (!floatBox.containingBlock()->style().marginTrim().contains(MarginTrimType::BlockEnd) || !floatBox.marginAfter())
+            continue;
+
+        // There are 3 different cases that need to be taken into consideration when trimming the
+        // block-end margin of a float:
+        // 1.) The float margin box extends past the deepest past of content (margin should be trimmed all the way)
+        // 2.) The end of the deepest piece of is in between the float's border box and marign box (marign should be trimmed up to the content)
+        // 3.) The deepest piece of content is below the float's margin box (no trimming occurs)
+        auto floatMarginAfterStartLocation = computeFloatMarginAfterStartLocationInBlockFormattingContext(floatBox);
+        if (floatMarginAfterStartLocation >= blockFormattingContextInFlowContentHeight)
+            trimMarginForFloat(*floatingObject, MarginTrimType::BlockEnd, floatBox.marginAfter(&style()));
+        else if (blockFormattingContextInFlowContentHeight > floatMarginAfterStartLocation && (blockFormattingContextInFlowContentHeight < floatMarginAfterStartLocation + floatBox.marginAfter()))
+            trimMarginForFloat(*floatingObject, MarginTrimType::BlockEnd, floatMarginAfterStartLocation + floatBox.marginAfter(&style()) - blockFormattingContextInFlowContentHeight);
+    }
+}
+
+bool RenderBlockFlow::shouldChildInlineMarginContributeToContainerIntrinsicSize(MarginTrimType marginSide, const RenderElement& child) const
+{
+    // For the purposes of computing a block container's intrinsic size, a float should not
+    // include its trimmed margins in its intrinsic size contributions
+    ASSERT(marginSide == MarginTrimType::InlineStart || marginSide == MarginTrimType::InlineEnd);
+    return !child.isFloating() || !style().marginTrim().contains(marginSide);
```

Comment:
This being a public function on RenderBlockFlow, what guarantees that the child's containing block is *this?

Reply:
Added this extra check to the assert to make sure that this condition holds